File: desktop/src-tauri/src/models/prompt.rs
//! Prompt Template Models
//!
//! Data structures for the prompt library feature.

use serde::{Deserialize, Serialize};

/// A prompt template in the library
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptTemplate {
    pub id: String,
    pub title: String,
    pub content: String,
    pub description: Option<String>,
    /// Category: "coding" | "writing" | "analysis" | "custom"
    pub category: String,
    /// Tags as a list of strings (stored as JSON in DB)
    pub tags: Vec<String>,
    /// Extracted {{variable}} names from content (stored as JSON in DB)
    pub variables: Vec<String>,
    /// Whether this is a built-in prompt (cannot be deleted)
    pub is_builtin: bool,
    /// Whether this prompt is pinned to the top
    pub is_pinned: bool,
    /// Number of times this prompt has been used
    pub use_count: u32,
    pub last_used_at: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

/// Request to create a new prompt template
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptCreateRequest {
    pub title: String,
    pub content: String,
    pub description: Option<String>,
    pub category: String,
    pub tags: Vec<String>,
    pub is_pinned: bool,
}

/// Request to update an existing prompt template
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptUpdateRequest {
    pub title: Option<String>,
    pub content: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub tags: Option<Vec<String>>,
    pub is_pinned: Option<bool>,
}

File: desktop/src/types/prompt.ts
/**
 * Prompt Template Types
 *
 * TypeScript types mirroring the Rust PromptTemplate model.
 */

export interface PromptTemplate {
  id: string;
  title: string;
  content: string;
  description: string | null;
  category: string;
  tags: string[];
  variables: string[];
  is_builtin: boolean;
  is_pinned: boolean;
  use_count: number;
  last_used_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface PromptCreateRequest {
  title: string;
  content: string;
  description: string | null;
  category: string;
  tags: string[];
  is_pinned: boolean;
}

export interface PromptUpdateRequest {
  title?: string;
  content?: string;
  description?: string | null;
  category?: string;
  tags?: string[];
  is_pinned?: boolean;
}

export interface CommandResponse<T> {
  success: boolean;
  data: T | null;
  error: string | null;
}

export const SYSTEM_PROMPT_CATEGORIES = ['coding', 'writing', 'analysis'] as const;
export const PROMPT_CATEGORIES = ['all', 'uncategorized', ...SYSTEM_PROMPT_CATEGORIES] as const;
export type PromptCategoryId = (typeof PROMPT_CATEGORIES)[number];

/** Map legacy/custom category markers to uncategorized. */
export function normalizePromptCategory(category: string | null | undefined): string {
  const normalized = (category ?? '').trim();
  return normalized.toLowerCase() === 'custom' ? '' : normalized;
}

/** Substitute {{variable}} placeholders in a template string */
export function substituteVariables(template: string, values: Record<string, string>): string {
  return template.replace(/\{\{(\w+)\}\}/g, (match, name) => {
    return values[name] !== undefined ? values[name] : match;
  });
}

/** Extract {{variable}} names from a template string */
export function extractVariables(template: string): string[] {
  const re = /\{\{(\w+)\}\}/g;
  const vars: string[] = [];
  let match;
  while ((match = re.exec(template)) !== null) {
    if (!vars.includes(match[1])) {
      vars.push(match[1]);
    }
  }
  return vars;
}
